### PR TITLE
feat: allow extension of default config

### PR DESCRIPTION
**Pull request by patrickhulce**
would address part of #1730, it's rather out there so comments and thoughts welcome on other possible approaches

usage

```
{
  extends: true,
  passes: [
    {passName: "defaultPass", gatherers: [MyCustomGatherer]
  ],
  audits: [MyCustomAudit]
}
```

### CODE REVIEW

File: lighthouse-core/config/config.js [hunk 1]
```diff
@@ -233,6 +233,27 @@ function expandArtifacts(artifacts) {
...
+    return value.concat(extension);
+  } else if (typeof extension === 'object') {
+    if (typeof value !== 'object') throw new TypeError(`Expected array but got ${typeof value}`);
+    Object.keys(extension).forEach(key => {
+      value[key] = merge(value[key], extension[key]);
+    });
+    return value;
+  } else {
+    return extension;
+  }
```
Comment (ebidel) [inline lighthouse-core/config/config.js]: nit: ditch `else` and just `return extension;`?

File: lighthouse-core/config/config.js [hunk 2]
```diff
@@ -262,6 +284,12 @@ class Config {
     if (Array.isArray(inputConfig.audits)) {
       configJSON.audits = Array.from(inputConfig.audits);
     }
+
+    // Extend the default config is specified
```
Comment (ebidel) [inline lighthouse-core/config/config.js]: // Extend the default config if specified.

File: lighthouse-core/config/config.js [hunk 3]
```diff
@@ -280,6 +308,27 @@ class Config {
     validatePasses(configJSON.passes, this._audits, this._configDir);
   }
 
+  static extendConfigJSON(baseJSON, extendJSON) {
```
Comment (ebidel) [inline lighthouse-core/config/config.js]: might want to add jsdocs, even though it's fairly obvious what everything is.

**Comment by ebidel** [comment]
This looks useful. I dreamed of similar things in https://github.com/GoogleChrome/lighthouse/issues/887. Our single gigantic config is being unwieldy. This will also help consumers of LH more easily create custom configs.

**Comment by ebidel** [comment]
Can you add documentation to the README? In the "creating custom configs" section.

File: lighthouse-core/config/config.js [hunk 4]
```diff
@@ -233,6 +233,27 @@ function expandArtifacts(artifacts) {
   return artifacts;
 }
 
+function merge(value, extension) {
```
Comment (paulirish) [inline lighthouse-core/config/config.js]: how about s/value/base/. communicates a little better which is being mutated, and in line with your `extendConfigJSON` below

File: lighthouse-core/config/config.js [hunk 5]
```diff
@@ -233,6 +233,27 @@ function expandArtifacts(artifacts) {
...
+  if (typeof value === 'undefined') {
+    return extension;
+  } else if (Array.isArray(extension)) {
+    if (!Array.isArray(value)) throw new TypeError(`Expected array but got ${typeof value}`);
+    return value.concat(extension);
+  } else if (typeof extension === 'object') {
+    if (typeof value !== 'object') throw new TypeError(`Expected array but got ${typeof value}`);
```
Comment (paulirish) [inline lighthouse-core/config/config.js]: array => object?

File: lighthouse-core/config/config.js [hunk 6]
```diff
@@ -280,6 +308,27 @@ class Config {
...
+          baseJSON.passes.push(pass);
+        } else {
+          merge(basePass, pass);
+        }
+      });
+
+      delete extendJSON.passes;
+    }
+
+    Object.keys(extendJSON).forEach(key => {
```
Comment (paulirish) [inline lighthouse-core/config/config.js]: couldn't these three lines be replaced with `merge(baseJSON, extendJSON)` ?

File: lighthouse-core/config/config.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: done

File: lighthouse-core/config/config.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: done

File: lighthouse-core/config/config.js (hunk 4 above)
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: done

File: lighthouse-core/config/config.js (hunk 5 above)
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: woops, done

File: lighthouse-core/config/config.js (hunk 6 above)
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: yep! artifact of earlier implementation, done

**Comment by patrickhulce** [comment]
@ebidel added a readme entry PTAL :)

**Pull request merged**